ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -0,0 +1,596 @@
+    Assert.assertEquals(RouterErrorCode.OperationTimedOut, routerException.getErrorCode());
+  }
+
+  /**
+   * Test the case where all requests time out within the NetworkClient.
+   * @throws Exception
+   */
+  @Test
+  public void testNetworkClientTimeoutAllFailure()
+      throws Exception {
+    doPut();
+    operationsCount.incrementAndGet();
+    GetBlobOperation op =
+        new GetBlobOperation(routerConfig, routerMetrics, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            null, operationCompleteCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      for (RequestInfo requestInfo : requestListToFill) {
+        ResponseInfo fakeResponse =
+            new ResponseInfo(requestInfo.getRequest(), NetworkClientErrorCode.NetworkError, null);
+        op.handleResponse(fakeResponse);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+      requestListToFill.clear();
+    }
+
+    // At this time requests would have been created for all replicas, as none of them were delivered,
+    // and cross-colo proxying is enabled by default.
+    Assert.assertEquals("Must have attempted sending requests to all replicas", replicasCount,
+        correlationIdToGetOperation.size());
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    RouterException routerException = (RouterException) op.getOperationException();
+    Assert.assertEquals(RouterErrorCode.OperationTimedOut, routerException.getErrorCode());
+  }
+
+  /**
+   * Test the case where every server returns Blob_Not_Found. All servers must have been contacted,
+   * due to cross-colo proxying.
+   * @throws Exception
+   */
+  @Test
+  public void testBlobNotFoundCase()
+      throws Exception {
+    doPut();
+    operationsCount.incrementAndGet();
+    GetBlobOperation op =
+        new GetBlobOperation(routerConfig, routerMetrics, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            null, operationCompleteCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+
+    for (MockServer server : mockServerLayout.getMockServers()) {
+      server.setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+    }
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+    }
+
+    Assert.assertEquals("Must have attempted sending requests to all replicas", replicasCount,
```

Comment:
this looks like its been repeated thrice. Perhaps calls for a utility function? Especially since the same test can be made for internal server errors, IO exceptions etc.